### ISSUE TITLE
'ChatGLMForConditionalGeneration' object has no attribute 'stream_chat'

### ISSUE THREAD

**Issue by reinforceli** [THUDM/GLM-4#365]
### System Info / 系統信息

windows10,A800,python3.10

### Information / 问题信息

- [ ] The official example scripts / 官方的示例脚本
- [ ] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

![image](https://github.com/user-attachments/assets/9e94168f-4d05-4648-8d9c-04e30860ed5a)
通Xinference启用模型，报错

### Expected behavior / 期待表现

能响应回答

**Comment by zRzRzRzRzRzRzR** [comment]
这个issue我们和xinfer沟通过了，这个函数并不是很规范，在我们的github发布的时候就已经全部替换成了generate这个标准的接口了

**Comment by cptbaker233** [comment]
您好, 请问您解决了吗, 我遇到和你一样的问题

**Comment by zzlTim** [comment]
你好，解决了吗 @cptbaker233 cptbaker233     @reinforceli 
@zRzRzRzRzRzRzR zRzRzRzRzRzRzR

**Comment by jhj033** [comment]
这个问题出在你下载的glm4-9b模型的~/.cache/modelscope/hub/ZhipuAI/glm-4-9b-chat/modeling_chatglm.py文件里没有stream_chat函数，我从modelscope中找了个带stream_chat函数的文件拷进去就work了

@torch.inference_mode()
def stream_chat(self, tokenizer, query: str, history: List[Dict] = None, role: str = "user",
                past_key_values=None, max_length: int = 8192, do_sample=True, top_p=0.8, temperature=0.8,
                logits_processor=None, return_past_key_values=False, **kwargs):
    if history is None:
        history = []
    if logits_processor is None:
        logits_processor = LogitsProcessorList()
    logits_processor.append(InvalidScoreLogitsProcessor())
    eos_token_id = [tokenizer.eos_token_id, tokenizer.convert_tokens_to_ids("<|user|>"),
                    tokenizer.convert_tokens_to_ids("<|observation|>")]
    gen_kwargs = {"max_length": max_length, "do_sample": do_sample, "top_p": top_p,
                  "temperature": temperature, "logits_processor": logits_processor, **kwargs}
    if past_key_values is None:
        inputs = tokenizer.apply_chat_template(history + [{"role": role, "content": query}],
                                               add_generation_prompt=True, tokenize=True, return_tensors="pt",
                                               return_dict=True)
    else:
        inputs = tokenizer.apply_chat_template([{"role": role, "content": query}], add_special_tokens=False,
                                               add_generation_prompt=True, tokenize=True, return_tensors="pt",
                                               return_dict=True)
    inputs = inputs.to(self.device)
    if past_key_values is not None:
        past_length = past_key_values[0][0].shape[2]
        inputs.position_ids += past_length
        attention_mask = inputs.attention_mask
        attention_mask = torch.cat((attention_mask.new_ones(1, past_length), attention_mask), dim=1)
        inputs['attention_mask'] = attention_mask
    history.append({"role": role, "content": query})
    for outputs in self.stream_generate(**inputs, past_key_values=past_key_values,
                                        eos_token_id=eos_token_id, return_past_key_values=return_past_key_values,
                                        **gen_kwargs):
        if return_past_key_values:
            outputs, past_key_values = outputs
        outputs = outputs.tolist()[0][len(inputs["input_ids"][0]):-1]
        response = tokenizer.decode(outputs)
        if response and response[-1] != "�":
            response, new_history = self.process_response(response, history)
            if return_past_key_values:
                yield response, new_history, past_key_values
            else:
                yield response, new_history

**Comment by zzlTim** [comment]
> 带stream_chat函数的文件拷进去就wo

刨根问底，厉害

**Comment by DentistCode** [comment]
> 这个问题出在你下载的glm4-9b模型的~/.cache/modelscope/hub/ZhipuAI/glm-4-9b-chat/modeling_chatglm.py文件里没有stream_chat函数，我从modelscope中找了个带stream_chat函数的文件拷进去就work了
> 
> ```
> @torch.inference_mode()
> def stream_chat(self, tokenizer, query: str, history: List[Dict] = None, role: str = "user",
>                 past_key_values=None, max_length: int = 8192, do_sample=True, top_p=0.8, temperature=0.8,
>                 logits_processor=None, return_past_key_values=False, **kwargs):
>     if history is None:
>         history = []
>     if logits_processor is None:
>         logits_processor = LogitsProcessorList()
>     logits_processor.append(InvalidScoreLogitsProcessor())
>     eos_token_id = [tokenizer.eos_token_id, tokenizer.convert_tokens_to_ids("<|user|>"),
>                     tokenizer.convert_tokens_to_ids("<|observation|>")]
>     gen_kwargs = {"max_length": max_length, "do_sample": do_sample, "top_p": top_p,
>                   "temperature": temperature, "logits_processor": logits_processor, **kwargs}
>     if past_key_values is None:
>         inputs = tokenizer.apply_chat_template(history + [{"role": role, "content": query}],
>                                                add_generation_prompt=True, tokenize=True, return_tensors="pt",
>                                                return_dict=True)
>     else:
>         inputs = tokenizer.apply_chat_template([{"role": role, "content": query}], add_special_tokens=False,
>                                                add_generation_prompt=True, tokenize=True, return_tensors="pt",
>                                                return_dict=True)
>     inputs = inputs.to(self.device)
>     if past_key_values is not None:
>         past_length = past_key_values[0][0].shape[2]
>         inputs.position_ids += past_length
>         attention_mask = inputs.attention_mask
>         attention_mask = torch.cat((attention_mask.new_ones(1, past_length), attention_mask), dim=1)
>         inputs['attention_mask'] = attention_mask
>     history.append({"role": role, "content": query})
>     for outputs in self.stream_generate(**inputs, past_key_values=past_key_values,
>                                         eos_token_id=eos_token_id, return_past_key_values=return_past_key_values,
>                                         **gen_kwargs):
>         if return_past_key_values:
>             outputs, past_key_values = outputs
>         outputs = outputs.tolist()[0][len(inputs["input_ids"][0]):-1]
>         response = tokenizer.decode(outputs)
>         if response and response[-1] != "�":
>             response, new_history = self.process_response(response, history)
>             if return_past_key_values:
>                 yield response, new_history, past_key_values
>             else:
>                 yield response, new_history
> ```
@jhj033 哥们方便提供一下完整代码么

目前使用了这段代码，参考 THUDM/glm-4-9b-chat-1m 的 [modeling_chatglm.py](https://huggingface.co/THUDM/glm-4-9b-chat-1m/blob/main/modeling_chatglm.py)，缺什么补什么，逐步报错，最终还是无果：
'GenerationConfig' object has no attribute '_eos_token_tensor’

同时尝试了把1m的代码直接给原版chat用，但是报同样的错



报错过程有：
* 'ChatGLMForConditionalGeneration' object has no attribute 'stream_chat’
* 补上process_response
* 'ChatGLMForConditionalGeneration' object has no attribute 'stream_generate’
* Callable name 'Callable' is not defined
  * from typing import Optional, Tuple, Union, List, Callable, Dict, Any 
* GenerationConfig' object has no attribute '_eos_token_tensor’

**Comment by DentistCode** [comment]
终于找到了个曲线救国的方案，直接加这里路径里就行：
![Snipaste_2024-08-11_20-14-42](https://github.com/user-attachments/assets/9947b328-057f-42e3-8b93-49e8ccb508d0)

**Comment by lergliu** [comment]
> 终于找到了个曲线救国的方案，直接加这里路径里就行： ![Snipaste_2024-08-11_20-14-42](https://private-user-images.githubusercontent.com/86056063/356873167-9947b328-057f-42e3-8b93-49e8ccb508d0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NDIyMzAsIm5iZiI6MTcyMzQ0MTkzMCwicGF0aCI6Ii84NjA1NjA2My8zNTY4NzMxNjctOTk0N2IzMjgtMDU3Zi00MmUzLThiOTMtNDllOGNjYjUwOGQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEyVDA1NTIxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViZmVjNTczNDcxYWIxMDBlM2RjOGJmNjJmNGYxODk2Zjc3OWI1MDI5YTQ1Yjg2NDU4MjE0ZDEwNzFiYzJhMGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.K-gB5VybhuBOAP5giIqandxkDhgArToQYqXA8qNdIRs)

请问您的modeling_chatglm.py代码是用的最初的还是您自行逐步补充过的？我这边加了路径，还是报GenerationConfig' object has no attribute '_eos_token_tensor’

**Comment by sbabybird** [comment]
https://modelscope.cn/models/dash-infer/glm-4-9b-chat-DI/file/view/master?fileName=modeling_chatglm.py&status=1

我使用魔塔社区里的这个版本中的 modeling_chatglm.py  直接替换掉原来的 modeling_chatglm.py 文件，就好使了。

**Comment by daydayfantasy** [comment]
> https://modelscope.cn/models/dash-infer/glm-4-9b-chat-DI/file/view/master?fileName=modeling_chatglm.py&status=1
> 
> 我使用魔塔社区里的这个版本中的 modeling_chatglm.py 直接替换掉原来的 modeling_chatglm.py 文件，就好使了。

这个方案会报
